analyzer/protocol/ldap/asn1.spicy

```diff
@@ -0,0 +1,339 @@
+  unused_bits: uint8;
+  value_bits: bytes &size=(len - 1);
+
+  # TODO - constructed form
+  # `bytes` needs << and >> support before we can implement complex bitstrings
+};
+
+#- ASN.1 octet string ---------------------------------------------------------
+# https://www.obj-sys.com/asn1tutorial/node10.html
+
+type ASN1OctetString = unit(len: uint64, constructed: bool) {
+  var len: uint64;
+  value: bytes &size = len;
+
+  # TODO - constructed form
+
+  on %done {
+    self.len = len;
+  }
+};
+
+#- ASN.1 various string types -------------------------------------------------
+# https://www.obj-sys.com/asn1tutorial/node124.html
+
+type ASN1String = unit(tag: ASN1Tag, len: uint64) {
+  var value: string = "";
+
+  octetstring: ASN1OctetString(len, tag.constructed) {
+
+    if ((tag.tpe == ASN1Type::PrintableString) ||
+        (tag.tpe == ASN1Type::GeneralizedTime) ||
+        (tag.tpe == ASN1Type::UTCTime)) {
+      self.value = self.octetstring.value.decode(hilti::Charset::ASCII);
+
+    } else if (tag.tpe == ASN1Type::UTF8String) {
+      self.value = self.octetstring.value.decode(hilti::Charset::UTF8);
+
+    } else {
+      # this would include ASN1Type::BMPString and ASN1Type::UniversalString because
+      # hilti::Charset::UTF16BE and hilti::Charset::UTF32BE don't exist any more
+      throw "Unsupported ASN1String type";
+    }
+  }
+
+};
+
+#- ASN.1 OID ------------------------------------------------------------------
+# https://www.obj-sys.com/asn1tutorial/node124.html
+
+type oidnibble = unit {
+  data : bitfield(8) {
+    num: 0..6;
+    more: 7;
+  };
+};
+
+type ASN1ObjectIdentifier = unit(len: uint64) {
+  var oid: vector<uint64>;
+  var temp: uint64;
+  var oidstring: string;
+
+  first: uint8 if ( len >= 1 ) {
+    self.temp = self.first / 40;
+    self.oid.push_back( self.temp );
+    self.oidstring = "%d" % (self.temp);
+    self.temp = self.first % 40;
+    self.oid.push_back( self.temp );
+    self.oidstring = self.oidstring + ".%d" % (self.temp);
+    self.temp = 0;
+  }
+
+  sublist: oidnibble[len - 1] foreach {
+    self.temp = ( self.temp<<7 ) | $$.data.num;
+    if ( $$.data.more != 1 ) {
+      self.oid.push_back(self.temp);
+      self.oidstring = self.oidstring + ".%d" % (self.temp);
+      self.temp = 0;
+    }
+  }
+
+};
+
+
+#- ASN.1 message header (tag + length information) ----------------------------
+
+public type ASN1Header = unit {
+  tag: ASN1Tag;
+
+  var total_length: uint64;
+
+  len: LengthType {
+    self.total_length = self.len.len + self.tag.len + self.len.tag_len;
+  }
+};
+
+#- ASN.1 message body ---------------------------------------------------------
+
+public type ASN1Body = unit(head: ASN1Header, recursive: bool, depth: uint32) {
+  switch ( head.tag.tpe ) {
+
+    ASN1Type::BOOLEAN -> bool_value: uint8 &convert=cast<bool>($$) {
+      if (head.len.len != 1) {
+        throw "ASN1Type::BOOLEAN detected with length != 1";
+      }
+    }
```

Comment:
If your check expresses a protocol invariant (i.e., we'd only ever hit it for invalid input), let's instead express it with a `&requires` attribute, e.g.,
```suggestion
    ASN1Type::BOOLEAN -> bool_value: uint8 &convert=cast<bool>($$) &requires=head.len.len==1;
    }
```